Replace empty string literals in Marshal.cs

tudoranghelina91 opened the pull request:
Replace emtpy string literals exception messages with null in Marshal.cs

Fix #68347

Commit:
Replace emtpy string literals in Marshal.cs

Replace emtpy string literals exception messages with null in Marshal.cs

Fix #68347

tudoranghelina91 commented:
@dotnet-policy-service agree

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.cs:
```diff
@@ -798,7 +798,7 @@ public static IntPtr GetHINSTANCE(Module m)
                     return new System.NotImplementedException();
                 //case HResults.E_POINTER:
                 case HResults.RO_E_CLOSED:
```
```suggestion
                case HResults.COR_E_OBJECTDISPOSED:
                case HResults.RO_E_CLOSED:
```
We can share the implementation for both cases.

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.cs:
```diff
@@ -739,7 +739,7 @@ public static IntPtr GetHINSTANCE(Module m)
                 case HResults.E_POINTER:
                     return new System.NullReferenceException();
                 case HResults.COR_E_OBJECTDISPOSED:
```
```suggestion
```

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.cs:
```diff
@@ -739,7 +739,7 @@ public static IntPtr GetHINSTANCE(Module m)
                     return new System.NullReferenceException();
                 case HResults.COR_E_OBJECTDISPOSED:
-                    return new System.ObjectDisposedException("");
+                    return new System.ObjectDisposedException(null);
```
```suggestion
```

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.cs:
```diff
@@ -824,7 +824,7 @@ public static IntPtr GetHINSTANCE(Module m)
                 case HResults.TYPE_E_TYPEMISMATCH:
                 case HResults.CO_E_NOTINITIALIZED:
                 case HResults.RPC_E_CHANGED_MODE:
```
Same here. The whole block of cases can be deleted. The `default:` case is going to take care of this just fine. (github does not allow me to make this as a suggestion.)

Commit:
Replace emtpy string literals exception messages with null in Marshal.cs

Replace emtpy string literals exception messages with null in Marshal.cs
Remove redundant switches in Marshal.cs
Group together switches for ObjectDisposedException

Fix #68347

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.cs:
```diff
@@ -798,33 +796,12 @@ public static IntPtr GetHINSTANCE(Module m)
-                case HResults.COR_E_CONTEXTMARSHAL:
+                case HResults.COR_E_OBJECTDISPOSED:
+                    return new System.ObjectDisposedException(null);
                 //case HResults.COR_E_HOSTPROTECTION:
```
```suggestion
```

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.cs:
```diff
@@ -798,33 +796,12 @@ public static IntPtr GetHINSTANCE(Module m)
-                case HResults.COR_E_MISSINGSATELLITEASSEMBLY:
-                case HResults.COR_E_SAFEARRAYRANKMISMATCH:
-                case HResults.COR_E_SAFEARRAYTYPEMISMATCH:
                 //case HResults.COR_E_SAFEHANDLEMISSINGATTRIBUTE:
                 //case HResults.COR_E_SEMAPHOREFULL:
                 //case HResults.COR_E_THREADSTOP:
```
```suggestion
```

jkotas committed:
Apply suggestions from code review

jkotas reviewed (approved):
Thanks!

danmoseley commented:
@jkotas  would we be open to API proposal to expose the parameterless ctor - that would tidy this up?

jkotas commented:
We want to encourage callers to pass in the extra information about the object that triggered the exception. The situations where this extra information is not available are extremely rare. I do not see a problem with callers having to pass in `null` explicitly in these rare cases.

jkotas commented:
There are no observable public API behavior changes. The ObjectName property getter [replaces `null` with empty string](https://github.com/dotnet/runtime/blob/a54a823ece1094dd05b7380614bd43566834a8f9/src/libraries/System.Private.CoreLib/src/System/ObjectDisposedException.cs#L99) and so the ObjectName property is going to return the empty string both before and after your change.

If there were observable public API behavior changes, it is unlikely we would accept the PR.

> Also, I see there are some failing checks in the pipeline

The failures are known issues. "Build Analysis" has links to github issues that are tracking these failures.
